### Setup

In [ ]:
#run setup python file
!python setup.py

import os
cwd = os.getcwd()
CODE_FOLDER = "/zeeguu-api/"
CODE_ROOT_FOLDER = cwd + "/zeeguu-api/"


from git import Repo
if not os.path.exists(CODE_ROOT_FOLDER):
    Repo.clone_from("https://github.com/zeeguu/api", CODE_ROOT_FOLDER)


# Count of lines of code in the project

In [12]:
# Install cloc if not already installed
!{sys.executable} -m pip install cloc --quiet

zsh:1: parse error near `-m'


In [13]:
#Run cloc to get the count of lines of code in the project we just downloaded 
!cloc .{CODE_FOLDER}


     859 text files.
     813 unique files.                                          
      54 files ignored.

github.com/AlDanial/cloc v 2.08  T=1.15 s (709.5 files/s, 385280.2 lines/s)
--------------------------------------------------------------------------------
Language                      files          blank        comment           code
--------------------------------------------------------------------------------
Text                             24          15585              0         298278
Python                          549          14464          14776          51064
HTML                             15           5734            773          21969
JSON                             20              0              0           4281
Markdown                         36           1372              5           3923
SQL                             136            424            498           2427
CSV                               3              0              0            462
Bou

From the cloc output above we can see that the project has 383,278 lines of code (excluding blank lines and comments). Out of these, 51,047 lines are Python code, which is the main programming language used in the project. The project also contains a significant amount of text files (298,278 lines), HTML (21,969 lines), and JSON (4,281 lines). 
The large amount of text files is likely due to the presence of data files, documentation, or other non-code resources in the project, which we will probably want to exclude from our architectural recovery analysis. 
We already know the project is zeeguu api and that the website frontend has its own repository, so the HTML files are likely to be templates or static files used by the API, and the JSON files are likely to be configuration files or data files used by the API. For the purpose of architectural recovery, we will focus on the Python code, as it is the main programming language used in the project and is likely to contain the core logic and structure of the application.

Other observations:
- The project contains a significant amount of Markdown files (3,923 lines), which are likely to be documentation files. These can provide valuable insights into the architecture and design decisions of the project, so we may want to review them as part of our architectural recovery process.
- The project also contains SQL files (2,427 lines), which are likely to be database schema definitions, queries, or migration scripts. These can provide insights into the data model and how the application interacts with the database, which is an important aspect of the architecture.
- The presence of Dockerfiles (52 lines) suggests that the project may be containerized, which can have implications for the deployment architecture and how the application is structured to run in a containerized environment.

In [14]:
# Exclude obvious non-project code (e.g. virtual environments, site-packages, and __pycache__) to get a more accurate count of lines of code in the project:
!cloc --exclude-dir=venv,site-packages,__pycache__ .{CODE_FOLDER}

     859 text files.
     813 unique files.                                          
      54 files ignored.

github.com/AlDanial/cloc v 2.08  T=1.16 s (698.1 files/s, 379098.8 lines/s)
--------------------------------------------------------------------------------
Language                      files          blank        comment           code
--------------------------------------------------------------------------------
Text                             24          15585              0         298278
Python                          549          14464          14776          51064
HTML                             15           5734            773          21969
JSON                             20              0              0           4281
Markdown                         36           1372              5           3923
SQL                             136            424            498           2427
CSV                               3              0              0            462
Bou

## Note on excluding non-project code from cloc count:
The cloc command was just cloc .{CODE_FOLDER} which counts all files in the project, including non-code files. To get a more accurate count of lines of code in the project, we can exclude obvious non-project code (e.g. virtual environments, site-packages, and __pycache__) using the following command:
!cloc --exclude-dir=venv,site-packages,__pycache__ .{CODE_FOLDER}
Running this we actually gave the same count of lines of code (383,278) because the excluded directories do not contain any code files that were counted in the initial cloc output. This suggests that the initial count of lines of code is already focused on the relevant project files and does not include extraneous code from virtual environments or site-packages.

In [15]:
# Focus on python code only:

!cloc --include-lang=Python --exclude-dir=venv,site-packages,__pycache__ .{CODE_FOLDER}

     859 text files.
     813 unique files.                                          
     318 files ignored.

github.com/AlDanial/cloc v 2.08  T=0.33 s (1661.9 files/s, 243097.9 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                         549          14464          14776          51064
-------------------------------------------------------------------------------
SUM:                           549          14464          14776          51064
-------------------------------------------------------------------------------


The above, where we focused on Python code only, does not really give any new information, as the initial count of lines of code already showed that there are 51,047 lines of Python code in the project. However, it is always nice to increase the focus on the main problem instead of getting distracted by other files that are not relevant to our architectural recovery analysis. By focusing on Python code only, we can ensure that we are analyzing the core logic and structure of the application, which is likely to be contained in the Python files.

In [16]:
#We can also exclude test, legacy (old), and migrations folders to get an even more accurate count of lines of code in the project:
!cloc --include-lang=Python --exclude-dir=venv,site-packages,__pycache__,test,old,migrations .{CODE_FOLDER}

     583 text files.
     538 unique files.                                          
     149 files ignored.

github.com/AlDanial/cloc v 2.08  T=0.39 s (1122.0 files/s, 182213.3 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                         442          12593          13759          45432
-------------------------------------------------------------------------------
SUM:                           442          12593          13759          45432
-------------------------------------------------------------------------------


Excluding test, legacy (old), and migrations folders gives us a more accurate count of lines of code in the project, as these folders often contain code that is not part of the main application logic. For example, test code is used for testing the application and may not reflect the actual architecture of the application. Legacy code may be old code that is no longer used or maintained, and migrations are typically scripts for database schema changes that may not be relevant to the current architecture. By excluding these folders, we can focus our analysis on the active and relevant codebase, which will provide a clearer picture of the architecture of the application.

All the above cloc commands are good to give an overview of the project and to get a sense of the size and composition of the codebase. However, they still have limitations in terms of providing insights into the architecture of the application. For example, they do not provide information about the structure of the code, the relationships between different components, or the design patterns used in the application. To gain a deeper understanding of the architecture, we will need to perform a more detailed analysis of the codebase, such as identifying modules, classes, and functions, and analyzing their interactions and dependencies. This will allow us to reconstruct the architecture of the application and identify key architectural decisions and patterns used in the design of the application.